### ISSUE TITLE
Evaluation on Reconstruction

### ISSUE THREAD

**Issue by Sissi-lu** [rmurai0610/MASt3R-SLAM#84]
Hi, 

Thank you for your excellent work in combining the MASt3R into the SLAM. I tried to reproduce the paper's results and use this model to see whether it is acceptable in a medical SLAM scenario. And I found some questions as follows:

1、I did not find the evaluation metrics for the reconstruction result (i.e., accuracy, completion, and chamber). I also tried to write the eval result, but the GT point cloud is too large if I generate it from the depth corresponding to the RGB images (7-scenes dataset). Could you publish this part of the code or give the instructions?

2、I noticed that not all the folders contain point clouds (ply) in EuRoC. So, did you evaluate only the scenes with point clouds?

3、I found it suitable for doing a slam in a gastrointestinal tract dataset. However, MASt3R did not see this kind of data. Why can it achieve this?

Many thanks for considering my request.

**Comment by rmurai0610** [comment]
Hi, we use only the depth from frames which were registered as keyframe to create the GT pointcloud. 
Similarly for EuRoC, we prune out the regions which are not observed by any of the keyframes.
We evaluate only on the EuRoC Vicon room, and all of the vicon rooms sequences has pointclouds.
I am not sure about the gastrointestinal tract dataset. MASt3R seems to generalise well beyond training data (e.g. it can more or less reconstruction static human, even though it's never seen one in the dataset) so I assume something similar is happening here. 

Hope these help!

**Comment by Sissi-lu** [comment]
Thank you once again for answering!

**Comment by Sissi-lu** [comment]
Hi,

Sorry to bother you once again. I still have some problems with reconstruction evaluation.

For reproducing reconstruction evaluation on 7 scenes, I followed your suggestion to use the depth from frames that were registered as keyframes to create the GT point cloud. And I found that the results using MASt3R-SLAM are higher than in the papers, as shown below:

<img width="408" alt="Image" src="https://github.com/user-attachments/assets/8581b256-708f-4f73-86f1-114a00d8c5ba" />

This is the evaluation code I wrote, as I did not find the reconstruction evaluation code in your repository:

`import argparse
import os.path
from natsort import natsorted
import yaml
import numpy as np
import open3d as o3d
import cv2
import torch
from mast3r_slam.dataloader import Intrinsics

def quaternion_to_rotation_matrix(qx, qy, qz, qw):
    """
    Convert a quaternion (qx, qy, qz, qw) to a 3x3 rotation matrix.
    Assumes quaternion is normalized (qx^2 + qy^2 + qz^2 + qw^2 = 1).

    Args:
        qx, qy, qz: Vector components of the quaternion.
        qw: Scalar component of the quaternion.

    Returns:
        R: 3x3 rotation matrix (numpy array).
    """
    # Ensure the quaternion is normalized
    norm = np.sqrt(qx ** 2 + qy ** 2 + qz ** 2 + qw ** 2)
    if not np.isclose(norm, 1.0, atol=1e-6):
        qx, qy, qz, qw = qx / norm, qy / norm, qz / norm, qw / norm

    # Compute the rotation matrix
    R = np.zeros((3, 3))
    R[0, 0] = 1 - 2 * qy ** 2 - 2 * qz ** 2
    R[0, 1] = 2 * qx * qy - 2 * qz * qw
    R[0, 2] = 2 * qx * qz + 2 * qy * qw
    R[1, 0] = 2 * qx * qy + 2 * qz * qw
    R[1, 1] = 1 - 2 * qx ** 2 - 2 * qz ** 2
    R[1, 2] = 2 * qy * qz - 2 * qx * qw
    R[2, 0] = 2 * qx * qz - 2 * qy * qw
    R[2, 1] = 2 * qy * qz + 2 * qx * qw
    R[2, 2] = 1 - 2 * qx ** 2 - 2 * qy ** 2

    return R


def pose_to_transformation_matrix(quat):
    """
    Convert a pose (tx, ty, tz, qx, qy, qz, qw) to a 4x4 transformation matrix.

    Args:
        tx, ty, tz: Translation components.
        qx, qy, qz, qw: Quaternion components (qw is the scalar part).

    Returns:
        T: 4x4 homogeneous transformation matrix (numpy array).
    """

    tx, ty, tz, qx, qy, qz, qw = quat
    # Get the 3x3 rotation matrix from the quaternion
    R = quaternion_to_rotation_matrix(qx, qy, qz, qw)

    # Create the 4x4 transformation matrix
    T = np.eye(4)  # Initialize 4x4 identity matrix
    T[0:3, 0:3] = R  # Set the top-left 3x3 block to the rotation matrix
    T[0:3, 3] = [tx, ty, tz]  # Set the translation vector in the last column

    return T

def undistort_image(image, intrinsics, dist_coeffs):
    """
    Undistort an image (depth or RGB) using camera intrinsics and distortion coefficients.

    Args:
        image (np.ndarray): Input image (depth: HxW, RGB: HxWx3).
        intrinsics (dict): Camera intrinsics with keys 'fx', 'fy', 'cx', 'cy'.
        dist_coeffs (np.ndarray): Distortion coefficients [k1, k2, p1, p2, k3].

    Returns:
        np.ndarray: Undistorted image.
    """
    # Create camera matrix
    camera_matrix = np.array([
        [intrinsics['fx'], 0, intrinsics['cx']],
        [0, intrinsics['fy'], intrinsics['cy']],
        [0, 0, 1]
    ])

    # OpenCV expects distortion coefficients in the order [k1, k2, p1, p2, k3]
    dist_coeffs = np.array(dist_coeffs)

    # Get image dimensions
    h, w = image.shape[:2]

    # Compute the undistortion map
    new_camera_matrix, _ = cv2.getOptimalNewCameraMatrix(camera_matrix, dist_coeffs, (w, h), 1, (w, h))
    mapx, mapy = cv2.initUndistortRectifyMap(camera_matrix, dist_coeffs, None, new_camera_matrix, (w, h), cv2.CV_32FC1)

    # Remap the image to undistort it
    undistorted_image = cv2.remap(image, mapx, mapy, interpolation=cv2.INTER_LINEAR)

    # Update intrinsics with the new camera matrix (if needed for further processing)
    intrinsics['fx'] = new_camera_matrix[0, 0]
    intrinsics['fy'] = new_camera_matrix[1, 1]
    intrinsics['cx'] = new_camera_matrix[0, 2]
    intrinsics['cy'] = new_camera_matrix[1, 2]

    return undistorted_image


def depth_to_point_cloud(depth_images, intrinsics, poes, dist_coeffs=None, rbg_images=None):
    """
    Convert a depth image to a point cloud using camera intrinsics, pose, and optional distortion correction.

    Args:
        depth_img (np.ndarray): Depth image (N, H, W).
        intrinsics (dict): Camera intrinsics with keys 'fx', 'fy', 'cx', 'cy'.
        pose (np.ndarray): Camera pose (N, 4x4 transformation matrix).
        dist_coeffs (np.ndarray, optional): Distortion coefficients [k1, k2, p1, p2, k3].
        rgb_img (np.ndarray, optional): RGB image (H, W, 3) for color.

    Returns:
        o3d.geometry.PointCloud: Point cloud.
    """
    points_world_list = []
    images_all = []
    for depth_img, pose, rgb_img in zip(depth_images, poes, rbg_images):
        # Copy intrinsics to avoid modifying the original
        intrinsics_copy = intrinsics.copy()

        # Undistort depth image if distortion coefficients are provided
        if dist_coeffs is not None:
            depth_img = undistort_image(depth_img, intrinsics_copy, dist_coeffs)

        # Undistort RGB image if provided
        if rgb_img is not None and dist_coeffs is not None:
            rgb_img = undistort_image(rgb_img, intrinsics_copy, dist_coeffs)

        height, width = depth_img.shape
        fx, fy, cx, cy = intrinsics_copy['fx'], intrinsics_copy['fy'], intrinsics_copy['cx'], intrinsics_copy['cy']

        # Create meshgrid of pixel coordinates
        u, v = np.meshgrid(np.arange(width), np.arange(height))
        z = depth_img  # Depth value at each pixel

        # Backproject to 3D camera coordinates
        x = (u - cx) * z / fx
        y = (v - cy) * z / fy
        points = np.stack([x, y, z], axis=-1).reshape(-1, 3)

        # Filter out invalid points (e.g., depth = 0)
        valid = (z.flatten() > 0)
        points = points[valid]

        # Transform points to world coordinates using the camera pose
        points_hom = np.hstack([points, np.ones((points.shape[0], 1))])  # (N, 4)
        points_world = (pose @ points_hom.T).T[:, :3]  # Apply pose transformation

        points_world_list.append(points_world)
        images_all.append((rgb_img.reshape(-1, 3)[valid]/255))

    images_all = np.concatenate(images_all, axis=0)
    pts_all = np.concatenate(points_world_list, axis=0)

    # Create Open3D point cloud
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(pts_all)
    pcd.colors = o3d.utility.Vector3dVector(images_all)

    return pcd

def merge_point_clouds(point_clouds):
    """
    Merge a list of point clouds into a single point cloud.

    Args:
        point_clouds (list of o3d.geometry.PointCloud): List of point clouds.

    Returns:
        o3d.geometry.PointCloud: Merged point cloud.
    """
    merged_pcd = o3d.geometry.PointCloud()
    points = []
    for pcd in point_clouds:
        points.append(np.asarray(pcd.points))
    merged_points = np.vstack(points)
    merged_pcd.points = o3d.utility.Vector3dVector(merged_points)
    return merged_pcd


def evaluate_reconstruction(pred_pcd, gt_pcds, threshold=0.5, device='cuda'):
     """
     Evaluate reconstruction using RMSE for Accuracy, Completion, and Chamfer Distance on GPU.

     Args:
     pred_pcd (o3d.geometry.PointCloud): Predicted point cloud.
     gt_pcds (list of o3d.geometry.PointCloud): List of ground truth point clouds.
     threshold (float): Maximum distance threshold (default: 0.5m).
     device (str): Device to run computations on ('cuda' or 'cpu').

     Returns:
     dict: Metrics including Accuracy (RMSE), Completion (RMSE), and Chamfer Distance.
     """
     # Check if GPU is available
     if device == 'cuda' and not torch.cuda.is_available():
         print("CUDA not available, falling back to CPU")
         device = 'cpu'

     # Merge GT point clouds
     merged_gt_pcd = merge_point_clouds(gt_pcds)

     # Convert to PyTorch tensors and move to device
     pred_points = torch.tensor(np.asarray(pred_pcd.points), dtype=torch.float32).to(device)
     gt_points = torch.tensor(np.asarray(merged_gt_pcd.points), dtype=torch.float32).to(device)

     # Compute pairwise distances using batch operations
     # pred_points: (N, 3), gt_points: (M, 3)
     # Compute squared distances: ||p - g||^2
     pred_points_exp = pred_points.unsqueeze(1) # (N, 1, 3)
     gt_points_exp = gt_points.unsqueeze(0).view(-1, 1, 3) # (1, M, 3)
     squared_dists = torch.sum((pred_points_exp - gt_points_exp) ** 2, dim=2) # (N, M)

     # Compute nearest neighbor distances
     # pred_to_gt_dists: min distance from each predicted point to any GT point
     pred_to_gt_dists = torch.sqrt(torch.min(squared_dists, dim=1)[0]) # (N,)
     # gt_to_pred_dists: min distance from each GT point to any predicted point
     gt_to_pred_dists = torch.sqrt(torch.min(squared_dists, dim=0)[0]) # (M,)

     # Apply threshold for Accuracy
     valid_pred_to_gt = pred_to_gt_dists < threshold
     if valid_pred_to_gt.sum() > 0:
        accuracy_rmse = torch.sqrt(torch.mean(pred_to_gt_dists[valid_pred_to_gt] ** 2))
     else:
         accuracy_rmse = torch.tensor(threshold, device=device)
     print(f"Accuracy (RMSE): {accuracy_rmse.item():.4f}")

     # Apply threshold for Completion
     valid_gt_to_pred = gt_to_pred_dists < threshold
     if valid_gt_to_pred.sum() > 0:
         completion_rmse = torch.sqrt(torch.mean(gt_to_pred_dists[valid_gt_to_pred] ** 2))
     else:
        completion_rmse = torch.tensor(threshold, device=device)
     print(f"Completion (RMSE): {completion_rmse.item():.4f}")

     # Compute Chamfer Distance
     mean_pred_to_gt = pred_to_gt_dists[valid_pred_to_gt].mean() if valid_pred_to_gt.sum() > 0 else torch.tensor(threshold, device=device)
     mean_gt_to_pred = gt_to_pred_dists[valid_gt_to_pred].mean() if valid_gt_to_pred.sum() > 0 else torch.tensor(threshold, device=device)
     chamfer_dist = (mean_pred_to_gt + mean_gt_to_pred) / 2.0
     print(f"Chamfer Distance: {chamfer_dist.item():.4f}")

     # Return metrics as dictionary
     return {
         "Accuracy_RMSE": accuracy_rmse.item(),
         "Completion_RMSE": completion_rmse.item(),
         "Chamfer_Distance": chamfer_dist.item()
         }


# Example usage
if __name__ == "__main__":
    # Camera intrinsics (replace with your actual values)
    parser = argparse.ArgumentParser()
    parser.add_argument("--dataset-path", type=str)
    parser.add_argument("--no-calib", action="store_true")

    args = parser.parse_args()
    split_dataset_type = args.dataset_path.split("/")
    dataset_path = args.dataset_path
    datatype, data_name = split_dataset_type[-2], split_dataset_type[-1]
    no_calib = args.no_calib
    base_path = "/data_new/luxiaoxi/code_proj/MASt3R-SLAM"
    # Load the predicted .ply file
    if no_calib:
        wo_calib = "no_calib"
    else:
        wo_calib = "calib"

    W, H = 640, 480
    fx, fy, cx, cy = 585.0, 585.0, 320.0, 240.0
    intrinsics = dict()
    intrinsics['fx'] = fx
    intrinsics['fy'] = fy
    intrinsics['cx'] = cx
    intrinsics['cy'] = cy
    distortion = None
    logs_path = os.path.join(base_path, "logs/%s/%s/%s/keyframes/%s"%(datatype, wo_calib, data_name, data_name))
    assert os.path.exists(logs_path), "%s does not exist"%logs_path
    keyframes_num = [int(f.split('.')[0]) for f in os.listdir(logs_path) if f.endswith('png')]
    rgb_files = []
    for num in keyframes_num:
        rgb_files.append(os.path.join(dataset_path, "seq-01", "frame-%06d.color.png"%(num)))
    depth_files = natsorted(f.replace("color", "depth") for f in rgb_files)  # List of depth images (H, W)

    poses_txt = "../groundtruths/7-scenes/%s.txt"%split_dataset_type[-1]
    with open(poses_txt, "r") as f:
        poses_list = f.readlines()
    poses_quat = []
    for num in keyframes_num:
        pose = [float(d) for d in poses_list[num].split(' ')[1:]]
        poses_quat.append(pose)


    depth_images = []
    rgb_images = []
    poses = []
    for i in range(len(rgb_files)):  # Assume 3 frames for the sequence
        depth_img = cv2.imread(depth_files[i], cv2.IMREAD_GRAYSCALE)
        rgb_img = cv2.imread(rgb_files[i])
        quat = poses_quat[i]
        pose = pose_to_transformation_matrix(quat)

        depth_images.append(depth_img)
        rgb_images.append(rgb_img)
        poses.append(pose)

    # Convert depth images to point clouds
    # gt_points = []
    # for depth_img, pose, rgb_img in zip(depth_images, poses, rgb_images):
    #     pcd = depth_to_point_cloud(depth_img, intrinsics, pose, rgb_img=rgb_img)
    #     gt_points.append(pcd)

    gt_pcds = depth_to_point_cloud(depth_images, intrinsics, poses, rbg_images=rgb_images)

    ply_path = os.path.join(base_path, "logs/%s/%s/%s/%s.ply"%(datatype, wo_calib, data_name, data_name))
    assert os.path.exists(ply_path), "there is no path: %s"%(ply_path)
    pred_pcd = o3d.io.read_point_cloud(ply_path)

    o3d.visualization.draw_geometries([pred_pcd], window_name="Predicted Point Cloud")

    o3d.visualization.draw_geometries([gt_pcds], window_name="GT Point Cloud")




    # Evaluate
    metrics = evaluate_reconstruction(pred_pcd, gt_pcds, threshold=0.5)
    print("\nFinal Evaluation Metrics:")
    print(f"Accuracy (RMSE): {metrics['Accuracy_RMSE']:.4f}")
    print(f"Completion (RMSE): {metrics['Completion_RMSE']:.4f}")
    print(f"Chamfer Distance: {metrics['Chamfer_Distance']:.4f}")`

Is the code right or not? If it is somewhere wrong, please point it out. Thank you once again for considering my request!
